docs/architecture/adr-071-bank-v2.md

```diff
@@ -0,0 +1,102 @@
+
+These issues have underscored the need for a refactoring strategy that simplifies the module’s architecture and focuses on its most essential operations.
+
+Additionally, there is an overlap in functionality with a Token Factory module, which could be integrated to streamline oper.
+
+## Decision
+
+**Permission Tightening**: Access to the module can be restricted to selected denominations only, ensuring that it operates within designated boundaries and does not exceed its intended scope. Currently, the permissions allow all denoms, so this should be changed. Send restrictions functionality will be maintained.
+
+**Simplification of Logic**: The bank module will focus on core functionalities `Send`, `Mint`, and `Burn`. This refinement aims to streamline the architecture, enhancing both maintainability and performance.
+
+**Integration of Token Factory**: The Token Factory will be merged into the bank module. This consolidation of related functionalities aims to reduce redundancy and enhance coherence within the system. Migrations functions will be provided for migrating from Osmosis' Token Factory module to bank/v2.
+
+**Legacy Support**: A legacy wrapper will be implemented to ensure compatibility with about 90% of existing functions. This measure will facilitate a smooth transition while keeping older systems functional.
+
+**Callback Functions**: We propose to to integrate callbacks which are handled by an`x/account` and will handle the callback message type. This allows for  for customisable behaviour during minting, burning and transferring operations without complicating statemanagement.It also helps overwrite the default accounting mechanism by allowing to overwrite the balance method (this is particularly useful for rebasing tokens).
+
+Each denom will be an `x/account` and will implement `OnTransfer`, `OnBurn`, `OnMint` and `OnBalanceOf` hooks. The bank module will check if the denom is an x/account and implement the respective hook. Depending on the hook, th bank module will send a message to the denom account. From there the denom account will return the new balances of the accounts.
+
+We would have the following messages `MsgMint`, `MsgBurn`, `MsgOnTransfer`, `QueryBalance` that would be sent to the denom account. 
+
+How the `x/account` denom implementation would look is as per below:
+
+```go
+func (a Account) RegisterExecuteHandlers(r accountsd.Router) {
+accountsd.RegisterHandler(r, a.OnTransfer)
+accountsd.RegisterHAndler(r. a.Burn)
+}
+
+func (a Account) OnTransfer(ctx context.Context, msg MsgOnTransfer) (MsgOnTransferResponse, error) {
+    ... check if account KYCd...
+}
+```
+
+For example if alice sends `MsgSend{to: bob, amount: 100ATOM}` to bank, bank checks if `ATOM` is an `x/account` implementing the `OnTransfer` hook.
+
+If `ATOM` is an `x/account` implementing the `OnTransfer` hook then bank sends a `MsgTransfer` to the `ATOM` denom account. The `ATOM` denom account returns the new balances of alice and bob.
```

Comment:
Will there be new functionality available to allow the ATOM account to use the bank ledger balance recording? I.e. will it be able to check some stuff, then do the transfer as normal?

Reply:
This was discussed during the working group call. Yes. Such functionality will be available.